File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/RadioGroup.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp

/**
 * KinetixRadioGroup / KinetixRadioButton — mirror
 * `packages/ui/src/components/radio-group.tsx` (`grid gap-3` group,
 * 18px/2px-border circle items). Figma-literal 18px size and 10dp dot
 * (`size-2.5`) aren't on the shared `spacing_*` scale — hardcoded, same
 * reasoning as `KinetixCheckbox`'s box. Grouping/selection state is left to
 * the caller (a `selectedOption` + `onSelect` pattern), same division of
 * responsibility Radix's own `RadioGroup`/`RadioGroupItem` split has —
 * `KinetixRadioGroup` only supplies the layout + `selectableGroup()`
 * semantics, same as [KinetixCheckbox]'s nullable-callback convention for
 * a disabled/read-only item.
 */
@Composable
fun KinetixRadioGroup(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Column(
        modifier = modifier.selectableGroup(),
        verticalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_3)),
    ) {
        content()
    }
}

@Composable
fun KinetixRadioButton(
    selected: Boolean,
    onClick: (() -> Unit)?,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    isError: Boolean = false,
) {
    val colors = KinetixColorScheme.current
    val borderColor = when {
        isError -> colors.destructive
        selected -> colors.primary
        else -> colors.border
    }
    val dotColor = if (isError) colors.destructive else colors.primary

    Box(
        modifier = modifier
            .size(18.dp)
            .clip(CircleShape)
            .selectable(
                selected = selected,
                onClick = { onClick?.invoke() },
                enabled = enabled && onClick != null,
                role = Role.RadioButton,
            )
            .border(dimensionResource(R.dimen.border_width_focus), borderColor, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        if (selected) {
            Box(modifier = Modifier.size(10.dp).clip(CircleShape).background(dotColor))
        }
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/RadioGroup.swift
//
// RadioGroup.swift — KinetixRadioGroup / KinetixRadioButton.
//
// Mirror packages/ui/src/components/radio-group.tsx (`grid gap-3` group,
// 18pt / 2pt-border circle items, 10pt dot). Figma-literal 18 and 10
// aren't on the shared spacing scale — hardcoded, same as the Compose
// port. Selection state is the caller's (`isSelected` + `action`) — the
// group only supplies layout, same division as Radix's own
// Root / Item split.
//

import SwiftUI

public struct KinetixRadioGroup<Content: View>: View {
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) { content } // grid gap-3
    }
}

public struct KinetixRadioButton: View {
    @Environment(\.kinetixColors) private var colors
    @Environment(\.isEnabled) private var isEnabled

    private let isSelected: Bool
    private let isError: Bool
    private let action: () -> Void

    public init(isSelected: Bool, isError: Bool = false, action: @escaping () -> Void) {
        self.isSelected = isSelected
        self.isError = isError
        self.action = action
    }

    private var borderColor: Color {
        if isError { return colors.destructive }
        return isSelected ? colors.primary : colors.input
    }

    public var body: some View {
        Button(action: action) {
            Circle()
                .strokeBorder(borderColor, lineWidth: 2)
                .frame(width: 18, height: 18)
                .overlay {
                    if isSelected {
                        Circle()
                            .fill(isError ? colors.destructive : colors.primary)
                            .frame(width: 10, height: 10)
                    }
                }
        }
        .buttonStyle(.plain)
        .opacity(isEnabled ? 1 : 0.5)
        .accessibilityAddTraits(isSelected ? [.isSelected] : [])
    }
}
